adapters/sovrnXsp/sovrnXsp.go

```diff
@@ -0,0 +1,167 @@
+package sovrnXsp
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/prebid/prebid-server/v2/adapters"
+	"github.com/prebid/prebid-server/v2/config"
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+
+	"github.com/prebid/openrtb/v19/openrtb2"
+)
+
+type adapter struct {
+	Endpoint string
+}
+
+// bidExt.CreativeType values.
+const (
+	creativeTypeBanner int = 0
+	creativeTypeVideo      = 1
+	creativeTypeNative     = 2
+	creativeTypeAudio      = 3
+)
+
+// Bid response extension from XSP.
+type bidExt struct {
+	CreativeType int `json:"creative_type"`
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if request.App != nil {
+		appCopy := *request.App
+		if appCopy.Publisher == nil {
+			appCopy.Publisher = &openrtb2.Publisher{}
+		} else {
+			publisherCopy := *appCopy.Publisher
+			appCopy.Publisher = &publisherCopy
+		}
+		request.App = &appCopy
+	} else {
+		return nil, []error{&errortypes.BadInput{
+			Message: "non-app request",
+		}}
+	}
```

Comment:
PBS-GO core framework will return error if ` request.App` is nil. No need for adapter code to handle it separately

Reply:
Resolved [1].

[1] https://github.com/prebid/prebid-server/pull/3312/commits/ae2e1090252bee341795f2ccbf3dcb4cb3641422